ideascube/serveradmin/wifi.py

```diff
@@ -0,0 +1,183 @@
+        self.__device = device
+        self.__connection = self.__get_connection()
+
+    def __get_connection(self):
+        for connection in NMSettings.ListConnections():
+            settings = connection.GetSettings()
+
+            try:
+                if settings['802-11-wireless']['ssid'] == self.ssid:
+                    return connection
+
+            except KeyError:
+                # This is not a Wi-Fi connection
+                continue
+
+    def __new_connection(self, wifi_key=None):
+        settings = {
+            'connection': {
+                'autoconnect': True,
+                'id': self.ssid,
+                'type': '802-11-wireless',
+                'uuid': str(uuid4()),
+                },
+            '802-11-wireless': {
+                'mode': 'infrastructure',
+                'ssid': self.ssid,
+                },
+            'ipv4': {
+                'method': 'auto'
+                },
+            'ipv6': {
+                'method': 'auto'
+                }
+            }
+
+        if self.secure:
+            if not wifi_key:
+                raise ValueError(
+                    _("A key is required to connect to this network"))
+
+            settings.update({
+                '802-11-wireless-security': {
+                    'auth-alg': 'open',
+                    'key-mgmt': 'wpa-psk',
+                    'psk': wifi_key,
+                    }})
+
+        return NMSettings.AddConnection(settings)
+
+    def connect(self, wifi_key=None):
+        if self.__connection is None:
+            self.__connection = self.__new_connection(wifi_key=wifi_key)
+
+        else:
+            NetworkManager.ActivateConnection(
+                self.__connection, self.__device, "/")
+
+    @property
+    def connected(self):
+        if self.__connection is None:
+            return False
+
+        for active_connection in NetworkManager.ActiveConnections:
+            if (active_connection.Connection.GetSettings()
+                    == self.__connection.GetSettings()):
+                return True
+
+        return False
+
+    @property
+    def known(self):
+        return self.__connection is not None
+
+    @property
+    def secure(self):
+        return self.__access_point.WpaFlags != 0
+
+    @property
+    def ssid(self):
+        return self.__access_point.Ssid
+
+    @property
+    def strength(self):
+        return self.__access_point.Strength
+
+
+class KnownWifiConnection(object):
+    @classmethod
+    def all(cls):
+        result = []
+
+        if not NetworkManager.WirelessEnabled:
+            return result
+
+        for connection in NMSettings.ListConnections():
+            if '802-11-wireless' not in connection.GetSettings():
+                continue
+
+            result.append(cls(connection))
+
+        return sorted(result, key=attrgetter('ssid'))
+
+    def __init__(self, connection):
+        self.__connection = connection
```

Comment:
That `__connection` attribute is a NetworkManager object. That is, it is an implementation detail, and should not be considered public API.

Sure, I could use only one `_` and rely on convention. But does it actually matter whether we rely on convention or on name mangling?

All I think matters is that this should be a private API, so that if we change the underlying implementation (say we move from NetworkManager to something else) then the rest of the app (the API user) is completely unaffected.

If you feel strongly about using `_` instead of `__`, I'll change it.